presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java

```diff
@@ -7081,6 +7108,148 @@ public void testUnsupportedCsvTable()
+                "\\QIncorrect timestamp precision for timestamp(9); the configured precision is: " + HiveTimestampPrecision.MICROSECONDS);
+        assertQueryFails(
+                session,
+                "CREATE TABLE test_invalid_precision_timestamp(ts) AS SELECT TIMESTAMP '2001-02-03 11:22:33.123'",
+                "\\QIncorrect timestamp precision for timestamp(3); the configured precision is: " + HiveTimestampPrecision.MICROSECONDS);
+        assertQueryFails(
+                session,
+                "CREATE TABLE test_invalid_precision_timestamp (ts TIMESTAMP(3))",
+                "\\QIncorrect timestamp precision for timestamp(3); the configured precision is: " + HiveTimestampPrecision.MICROSECONDS);
+    }
+
+    @Test
+    public void testTimestampPrecisionInsert()
+    {
+        testWithAllStorageFormats(this::testTimestampPrecisionInsert);
+    }
+
+    private void testTimestampPrecisionInsert(Session session, HiveStorageFormat storageFormat)
+    {
+        if (storageFormat == HiveStorageFormat.AVRO) {
+            // Avro timestamps are stored with millisecond precision
+            return;
+        }
+
+        String createTable = "CREATE TABLE test_timestamp_precision (ts TIMESTAMP) WITH (format = '%s')";
+        @Language("SQL") String insert = "INSERT INTO test_timestamp_precision VALUES (TIMESTAMP '%s')";
+
+        testTimestampPrecisionWrites(
+                session,
+                (ts, precision) -> {
+                    assertUpdate("DROP TABLE IF EXISTS test_timestamp_precision");
+                    assertUpdate(format(createTable, storageFormat));
+                    assertUpdate(withTimestampPrecision(session, precision), format(insert, ts), 1);
+                });
+    }
+
+    @Test
+    public void testTimestampPrecisionCtas()
+    {
+        testWithAllStorageFormats(this::testTimestampPrecisionCtas);
+    }
+
+    private void testTimestampPrecisionCtas(Session session, HiveStorageFormat storageFormat)
+    {
+        if (storageFormat == HiveStorageFormat.AVRO) {
+            // Avro timestamps are stored with millisecond precision
+            return;
+        }
+
+        String createTableAs = "CREATE TABLE test_timestamp_precision WITH (format = '%s') AS SELECT TIMESTAMP '%s' ts";
+
+        testTimestampPrecisionWrites(
+                session,
+                (ts, precision) -> {
+                    assertUpdate("DROP TABLE IF EXISTS test_timestamp_precision");
+                    assertUpdate(withTimestampPrecision(session, precision), format(createTableAs, storageFormat, ts), 1);
+                });
+    }
+
+    private void testTimestampPrecisionWrites(Session session, BiConsumer<String, HiveTimestampPrecision> populateData)
+    {
+        populateData.accept("2019-02-03 18:30:00.123", HiveTimestampPrecision.MILLISECONDS);
+        @Language("SQL") String sql = "SELECT ts FROM test_timestamp_precision";
+        assertQuery(
```

Comment:
These would be easier to read without wrapping, since it will put the values next to each other
```java
assertQuery(withTimestampPrecision(session, HiveTimestampPrecision.MILLISECONDS), sql, "VALUES ('2019-02-03 18:30:00.123')");
assertQuery(withTimestampPrecision(session, HiveTimestampPrecision.MICROSECONDS), sql, "VALUES ('2019-02-03 18:30:00.123')");
assertQuery(withTimestampPrecision(session, HiveTimestampPrecision.NANOSECONDS), sql, "VALUES ('2019-02-03 18:30:00.123')");